Cache startup files

poechsel opened the pull request:
Most of the work when linking binaries is spend generating the `genfns` in the startup file (`camlapply`, `camltuple`). These can be generated once and reused later on.
This pr adds:
- `ocamlopt -gen-cached-startup bin/cached-startup -linscan` that will create a file `bin/cached-startup.o` containing most `genfns`
- An option `use-cached-startup <bin/cached-startup.o>` that will tell the linker to assume most `genfns` don't need to be generated and whose implementation will be in `bin/cached-startup.o`

Most of the possible `genfns` are put in this store - `-gen-cached-startup` takes ~1min and creates an `.o` file of around 5mb. When `-use-cached-startup` is used every `genfns` goes through a whitelist of generated function and gets generated only if not in the whitelist.

Overall this improves linking time by over 25% when linking large binaries.

gretay-js commented on ocaml/utils/clflags.mli:
```diff
@@ -199,6 +199,8 @@ val afl_instrument : bool ref
 val afl_inst_ratio : int ref
 val function_sections : bool ref
 val probes : bool ref
+val use_cached_startup : string option ref
```
These flags can be in `driver/flambda_backend_flags.ml` because they are only used in the backend.

gretay-js reviewed (commented):
I have a few small suggestions but my high level comment is that we can probably avoid `-gen-cached-startup` and the argument of `-use-cached-startup` if we put the generated file in the compiler distribution, somewhere in `lib/ocaml`.

poechsel commented on backend/cmm_helpers.ml:
```diff
@@ -2977,6 +2977,106 @@ module Generic_fns_tbl = struct
+        |> List.concat
+      in
+      let apply =
+        List.init 404 (fun n ->
```
I don't exactly know, but they show up in our codebase.
According to Leo the max arity limit doesn't apply for `caml_apply` and maybe we should replace one call to `caml_apply400` by three calls to `caml_apply128` and one to `caml_apply22`

poechsel commented on driver/optmaindriver.ml:
```diff
@@ -80,7 +80,13 @@ let main unix argv ppf ~flambda2 =
     if !Clflags.plugin then
       Compenv.fatal "-plugin is only supported up to OCaml 4.08.0";
-    begin try
+    if Option.is_some !Clflags.gen_cached_startup then begin
```
I've moved the generation of the cached startup to a separate tool

poechsel commented on backend/cmm_helpers.ml:
```diff
@@ -2958,14 +2958,14 @@ module Generic_fns_tbl = struct
     let t = make () in
-    add t fns;
+    add_uncached t fns;
     t
```
Good catch

poechsel commented on backend/asmlink.ml:
```diff
@@ -447,12 +452,51 @@ let link_shared unix ~ppf_dump objfiles output_name =
+  List.iter compile_phrase
+    (Cmm_helpers.emit_preallocated_blocks []
+      (Cmm_helpers.generic_functions false genfns)));
+  if !Clflags.output_complete_object then
+    force_linking_of_startup ~ppf_dump;
```
I removed it

poechsel commented on backend/cmm_helpers.ml:
```diff
@@ -2977,6 +2977,106 @@ module Generic_fns_tbl = struct
+        Cmx_format.
+          { curry_fun = List.filter (fun x -> is_curry x) (tuplify @ curry);
+            send_fun = List.filter (fun x -> is_send x) send;
+            apply_fun = List.filter (fun x -> is_apply x) apply
```
The source of truth of what gets in the cache are these filters. Generating the cached startup is mainly about trying enough candidates to cover all function that are actually getting cached, meaning that we might try too many candidates. The important part is to try enough of them

poechsel commented:
I've:
- Added a new flag `-cached-startup-path` to set the path to the cached startup file. By default the compiler will look at `bindir/cached-startup.o`
- Removed the cached startup from the compiler to a separate tool named `tools/generate_cached_startup.exe`

gretay-js commented on driver/flambda_backend_flags.ml:
```diff
@@ -354,5 +354,8 @@ let set_o3 () =
   { set_oclassic; set_o2; set_o3 }
 
+let use_cached_startup = ref false
+let cached_startup_path = ref (Filename.concat Config.bindir ("cached-startup" ^ Config.ext_obj))
```
I don't think we have any other `.o` files in bin directory, `lib/ocaml` may be a better place. Also I would rather have one flag: if the path is hardcoded in the compiler, then the flag can just be boolean. Otherwise, the flag should just have the path, but I don't feel strongly about it.

poechsel replied:
I don't agree with that - if we make the path fully hardcoded, then testing the change might be a bit harder. The other motivation of having two files is to shorten the command line invocation: instead of always passing `-use-xxx /some/path/bin/cached-generic-functions.o` we can now only pass `-use-xxx`

gretay-js reviewed (commented):
Thanks for the changes. I think `make install` needs to be updated to install the generated `.o` file. 

It would be nice to have a test that exercises the new flag as part of `make runtest`, even if it only refers to the precomputed generic functions. It may be tricky to contrive a small test that needs a generic function that is not already precomputed.

poechsel commented on backend/asmlink.ml:
```diff
@@ -447,12 +448,51 @@ let link_shared unix ~ppf_dump objfiles output_name =
     remove_file startup_obj
   )
 
+let make_cached_startup_functions unix  ~ppf_dump ~sourcefile_for_dwarf =
```
Done

poechsel commented on backend/asmlink.ml:
```diff
@@ -401,10 +406,6 @@ let make_shared_startup_file unix ~ppf_dump ~sourcefile_for_dwarf genfns units =
-  if !Clflags.output_complete_object then
-    force_linking_of_startup ~ppf_dump;
-  (* this is to force a reference to all units, otherwise the linker
-     might drop some of them (in case of libraries) *)
```
My bad, fixed

poechsel commented:
Generating the cached startup file takes around 1 minute. Do we really want to add that to `make install` by default?

gretay-js commented on driver/flambda_backend_args.ml:
```diff
@@ -495,6 +495,15 @@ let mk_no_dwarf_for_startup_file f =
+
+let mk_cached_generic_functions_path f =
+  "-cached-generic-functions-path", Arg.String f,
+  "<file>  Set the path of the cached startup (default to cache-startup.o)"
```
The default name does not match `Flambda_backend_flags.cached_generic_functions_path`.

poechsel replied:
Fixed in the next commit

mshinwell reviewed (approved):
Only driver/flambda_backend* and tools/dune
